scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,367 @@
+                with that dispatch_uid.
+
+            sender
+                The sender to which the receiver should respond. Must either be
+                of type Signal, or None to receive events from any sender.
+
+            weak
+                Whether to use weak references to the receiver. By default, the
+                module will attempt to use weak references to the receiver
+                objects. If this parameter is false, then strong references
+                will be used.
+
+            dispatch_uid
+                An identifier used to uniquely identify a particular instance
+                of a receiver. This will usually be a string, though it may be
+                anything hashable.
+        """
+
+        # If DEBUG is on, check that we got a good receiver
+        # if settings.DEBUG:
+        assert callable(receiver), "Signal receivers must be callable."
+
+        # Check for **kwargs
+        if not func_accepts_kwargs(receiver):
+            raise ValueError(
+                "Signal receivers must accept keyword arguments (**kwargs).")
+
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+
+        if weak:
+            ref = weakref.ref
+            receiver_object = receiver
+            # Check for bound methods
+            if hasattr(receiver, '__self__') and hasattr(receiver, '__func__'):
+                ref = WeakMethod
+                receiver_object = receiver.__self__
+            if six.PY3 and sys.version_info >= (3, 4):
+                receiver = ref(receiver)
+                weakref.finalize(receiver_object, self._remove_receiver)
+            else:
+                receiver = ref(receiver, self._remove_receiver)
+
+        with self.lock:
+            self._clear_dead_receivers()
+            for r_key, _ in self.receivers:
+                if r_key == lookup_key:
+                    break
+            else:
+                self.receivers.append((lookup_key, receiver))
+            self.sender_receivers_cache.clear()
+
+    def disconnect(self, receiver=None, sender=None, weak=None, dispatch_uid=None):
+        """
+        Disconnect receiver from sender for signal.
+
+        If weak references are used, disconnect need not be called. The receiver
+        will be remove from dispatch automatically.
+
+        Arguments:
+
+            receiver
+                The registered receiver to disconnect. May be none if
+                dispatch_uid is specified.
+
+            sender
+                The registered sender to disconnect
+
+            dispatch_uid
+                the unique identifier of the receiver to disconnect
+        """
+        if weak is not None:
+            warnings.warn("Passing `weak` to disconnect has no effect.",
+                          ScrapyDeprecationWarning, stacklevel=2)
```

Comment:
We can probably drop `weak` from the keyword arguments, no use in introducing already deprecated code. ;)